### PR TITLE
Refactor note commands

### PR DESCRIPTION
**Pull request by cutandpastey**
This PR refactors most of the core login of the plugin.  Depends on https://github.com/guardian/scribe-plugin-noting/pull/62 @robinedman @hmgibson23 

### QA
With this on I'd pretty much try everything.  Ive done it here locally and looks fine but would be good to get a second pair of eyes on it.

### CODE REVIEW

**Comment by rrees** [comment]
Could you fix the Travis build?

File: src/NoteCommandFactory.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+var emitter = require('./utils/emitter');
```
Comment (robinedman) [inline src/NoteCommandFactory.js]: The `NoteCommandFactory.js` and `NoteController.js` filenames are in camel case while all other names are separated by dashes. What convention should we use?

Comment (hmgibson23) [inline src/NoteCommandFactory.js]: I favour dashes for filenames personally...

On 14 January 2015 at 11:23, Robin Edman <notifications@github.com> wrote:

> In src/NoteCommandFactory.js
> <https://github.com/guardian/scribe-plugin-noting/pull/65#discussion-diff-22930689>
> :
>
> > @@ -0,0 +1,66 @@
> > +var emitter = require('./utils/emitter');
>
> The NoteCommandFactory.js and NoteController.js filenames are in camel
> case while all other names are separated by dashes. What convention should
> we use?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/scribe-plugin-noting/pull/65/files#r22930689>
> .
>

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

File: src/actions/noting/remove-note.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    errorHandle('Only a valid VFocus element can be passed to removeNote, you passed: %s', focus);
+  }
+
+  // We assume the caller knows there's only one marker.
+  var marker = findScribeMarkers(focus)[0];
+
+
+  // We can't use findEntireNote here since it'll sometimes give us the wrong result.
+  // See `findEntireNote` documentation. Instead we look the note up by its ID.
+  noteCache.set(focus);
```
Comment (robinedman) [inline src/actions/noting/remove-note.js]: Why do we cache focus here?

Comment (cutandpastey) [inline src/actions/noting/remove-note.js]: This was in the original code base.

File: src/actions/noting/remove-note.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  if (!isVFocus(focus)){
+    errorHandle('Only a valid VFocus element can be passed to removeNote, you passed: %s', focus);
+  }
+
+  // We assume the caller knows there's only one marker.
+  var marker = findScribeMarkers(focus)[0];
+
+
+  // We can't use findEntireNote here since it'll sometimes give us the wrong result.
```
Comment (robinedman) [inline src/actions/noting/remove-note.js]: This comment would be better placed above line 28 I think.

File: src/actions/noting/remove-note.js (hunk 2 above)
Comment (hmgibson23) [inline src/actions/noting/remove-note.js]: My understanding is that ```noteCache.set``` requires the focus to rebuild
the tree and cache it. So it's not technically caching the focus right?
Maybe the function should be renamed.

On 14 January 2015 at 11:29, Jon Parsons <notifications@github.com> wrote:

> In src/actions/noting/remove-note.js
> <https://github.com/guardian/scribe-plugin-noting/pull/65#discussion-diff-22930976>
> :
>
> > +
> > +// treeFocus: tree focus of tree containing two scribe markers
> > +// Note that we will mutate the tree.
> > +module.exports = function removeNote(focus){
> > +
> > +  if (!isVFocus(focus)){
> > +    errorHandle('Only a valid VFocus element can be passed to removeNote, you passed: %s', focus);
> > +  }
> > +
> > +  // We assume the caller knows there's only one marker.
> > +  var marker = findScribeMarkers(focus)[0];
> > +
> > +
> > +  // We can't use findEntireNote here since it'll sometimes give us the wrong result.
> > +  // See `findEntireNote` documentation. Instead we look the note up by its ID.
> > +  noteCache.set(focus);
>
> This was in the original code base.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/scribe-plugin-noting/pull/65/files#r22930976>
> .
>

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

File: src/actions/noting/remove-note.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  // See `findEntireNote` documentation. Instead we look the note up by its ID.
+  noteCache.set(focus);
+
+  var note = findParentNoteSegment(marker);
```
Comment (robinedman) [inline src/actions/noting/remove-note.js]: Picky now, but for consistency, could we call the variable `noteSegment`? (As a "note" should be an array of note segments)

File: src/utils/noting/find-last-note-segment.js [hunk 5]
```diff
@@ -1,18 +1,14 @@
...
 
   if (!isVFocus(focus)) {
     errorHandle('only a valid VFocus can be passed to findFirstNoteSegment, you passed: %s', focus);
   }
 
-  return _.last(
-    focus.takeWhile(isWithinNote).filter(isNoteSegment)
-  );
-
+  return focus.takeWhile(isWithinNote).filter(isNoteSegment).splice(-1)[0];
```
Comment (robinedman) [inline src/utils/noting/find-last-note-segment.js]: Never thought of doing it that way, but that's pretty neat! To use less of lodash I guess?

File: test/unit/actions/noting/create-note-from-selection.spec.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      h('em.scribe-marker'),
+      h('p', 'This is some text'),
+    ]);
+
+    tree = new VFocus(tree);
+    tree = createNoteFromSelection(tree);
+
+    expect(tree.next().vNode.tagName).not.to.equal('gu-note');
+    expect(tree.next().next().vNode.tagName).not.to.equal('gu-note');
+
```
Comment (robinedman) [inline test/unit/actions/noting/create-note-from-selection.spec.js]: I find it hard to follow how this test tests that the note only wraps the selected test.

Also, would it make sense to also check for the actual text?

File: test/unit/actions/noting/remove-part-of-note.spec.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+      new VText('some'),
+      h('em.scribe-marker'),
+      new VText('text')
+    ])
+    ]);
+
+    tree = new VFocus(tree);
+    tree = removePartOfNote(tree);
+
+    expect(JSON.stringify(tree).match(/gu-note/g).length).to.be.equal(3);
```
Comment (robinedman) [inline test/unit/actions/noting/remove-part-of-note.spec.js]: This checks that _a_ note is removed, not necessarily that it's the selected note that's removed, doesn't it?

File: src/actions/noting/create-note-from-selection.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  // Replace the nodes in the tree with the wrapped versions.
+  _.zip(toWrapAndReplace, wrappedTextNodes).forEach(focusAndReplacementVNode => {
+    var focus = focusAndReplacementVNode[0];
+    var replacementVNode = focusAndReplacementVNode[1];
+    focus.replace(replacementVNode);
+  });
+
+  // If we end up with an empty note a <BR> tag would be created. We have to do
+  // this before we remove the markers.
+  removeErroniousBrTags(focus);
```
Comment (hmgibson23) [inline src/actions/noting/create-note-from-selection.js]: This is erroneous

Comment (cutandpastey) [inline src/actions/noting/create-note-from-selection.js]: Damn, I typo'd it and my auto complete picked it up :(

File: test/unit/actions/noting/remove-part-of-note.spec.js (hunk 7 above)
Comment (cutandpastey) [inline test/unit/actions/noting/remove-part-of-note.spec.js]: There's only one note to begin with.

File: src/actions/noting/remove-erroneous-br-tags.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+  //find the previous and next note segment
+  var segments = [
+    marker.find(isNoteSegment, 'prev'),
+    marker.find(isNoteSegment)
+  ].filter(o => !!o);
+
+  // Replace/delete empty notes, and parents that might have become empty.
+  segments.map(segment => {
+    if (hasOnlyEmptyTexChildren(segment)){
```
Comment (hmgibson23) [inline src/actions/noting/remove-erroneous-br-tags.js]: Should be ```hasOnlyEmptyTextChildren``` I assume.

**Comment by robinedman** [comment]
:+1:
